tools/nightly_msrv.sh

```diff
@@ -0,0 +1,101 @@
+#! /bin/bash
+
+#
+# Ensure all necessary dependencies are installed
+#
+
+if ! command -v rustup >/dev/null 2>&1; then
+	echo -e "\e[34mINFO: rustup is not installed, attempting to install...\e[0m"
+	curl --proto '=https' --tlsv1.2 -sSf https://sh.rustup.rs | sh
+	# refresh environment to load rustup
+	source "$HOME"/.cargo/env
+fi
+
+# Determines if a passed version of rust is compatible with the evaluated crate
+function is_compatible() {
+	local version="$1"
+
+	# use rustup to install the version of rust
+	rustup install "nightly-$version" --profile default
+	rustup default "nightly-$version"
+
+	# check if the crate compiles with the version of rust
+	if cargo check --all; then
+		echo -e "\e[32mSUCCESS: Found compatible version: $version\e[0m"
+		return 0
+	else
+		echo -e "\e[31mINFO: Version $version is not compatible\e[0m"
+		return 1
+	fi
+}
+
+# copy the currently installed toolchain to restore later
+current_toolchain=$(rustup show active-toolchain | cut -d ' ' -f 1)
+
+# check for the declared edition of the crate
+edition=$(grep -m 1 "edition" Cargo.toml | cut -d '"' -f 2)
+# earliest date to check for compatibility
+start_date="$edition-01-01"
+msrv=$(date +%Y-%m-%d)
+
+# Calculate the end date for the binary search
+end_date="$msrv"
+
+# use binary search to find the earliest compatible version of rust
+while [[ "$start_date" != "$end_date" ]]; do
+	# Display the progress: ruled-out count / total count
+	echo -e "\e[36mINFO: Searching between nightly releases $start_date and $end_date\e[0m"
+
+	# Calculate the midpoint date
+	mid_date=$(date -d "$start_date + $((($(date -d "$end_date" +%s) - $(date -d "$start_date" +%s)) / 2)) seconds" +"%Y-%m-%d")
+
+	# check if the crate is compatible with the version of rust
+	if is_compatible "$mid_date"; then
+		# If compatible, update the end date for the next iteration
+		end_date="$mid_date"
+	else
+		# If not compatible, update the start date for the next iteration
+		start_date=$(date -d "$mid_date + 1 day" +"%Y-%m-%d")
+	fi
+done
```

Comment:
so... it seems that we are repeating ourselves by writing something similar to the https://github.com/foresterre/cargo-msrv you have mentioned? maybe we can call that directly, then we will not need to maintain this script within FRB. (if that package misses some features we need, we can PR to that package first - then, everyone - not only frb developers - will benefit from your code about MSRV!)

Reply:
The reasons I decided to implement this from scratch are:

a) https://github.com/foresterre/cargo-msrv does NOT support nightly (perhaps I'll make a PR there in the future) and

b) implementing it from scratch allows for find-and-replace to automate updating/adding versions to CI, README, book, etc.

Reply:
> a) [foresterre/cargo-msrv](https://github.com/foresterre/cargo-msrv) does NOT support nightly (perhaps I'll make a PR there in the future)

Yes I think so, that's why I hope you to PR there - then everyone in Rust ecosystem will benefit from your code about MSRV.

> b) implementing it from scratch allows for find-and-replace to automate updating/adding versions to CI, README, book, etc.

To have a good separate of concerns, it would be great to extract it to a function separate from determining MSRV (as is mentioned in another comment)

Reply:
> > a) [foresterre/cargo-msrv](https://github.com/foresterre/cargo-msrv) does NOT support nightly (perhaps I'll make a PR there in the future)
> 
> Yes I think so, that's why I hope you to PR there - then everyone in Rust ecosystem will benefit from your code about MSRV.

I whole-heartedly agree.

To be completely honest, I think that this PR should be closed, and once I/the dev team @ https://github.com/foresterre/cargo-msrv implement and release nightly MSRV checking, another different PR should be reviewed.

there are too many irrelevant commits in this PR and simply put, there is no reason.

I'll probably open a draft PR to track progress, but I'll mark as ready for merge and send a ping when we're ready.

Apologies my friend for the lack of communication prior to opening this PR